### ISSUE TITLE
serverPrefetch doesn't work as expected

### ISSUE THREAD

**Issue by DominusVilicus** [vuejs/vue#9906]
### Version
2.6.10

### Steps to reproduce
Suppose we have a component:
```js
export default {
    data(){
        return { foo: null}
    }
    async serverPrefetch(){
        this.foo = 'serverPrefetch'
    },
    created(){
        this.foo = 'created'
    },
}
```
`foo` is initially null, then `created()` updates it to `created` and this reflects in the html interpolitatiion, but when serverPrefetch() updates `this.foo` it is not reflected on the rendered html, despite being called afterwards.

### What is expected?
serverPrefetch() should reflect rendered and updated data variables when rendered on SSR

### What is actually happening?
serverPrefetch() doesn't update the rendered view with new data after being called

**Comment by DominusVilicus** [comment]
Changes to state/data in `serverPrefetch()` should retrigger a VNode re-render for the affected components *watching* the reactive data

**Comment by Akryum** [comment]
The component isn't rendered before all `serverPrefetch` lifecycle hooks are called. [See the passing test](https://github.com/vuejs/vue/blob/8082d2f910d963f14c151fb445e0fcc5c975cca9/test/ssr/ssr-string.spec.js#L1350).

Do you have a runnable reproduction?

**Comment by DominusVilicus** [comment]
It seems that it is sending the correct content over on the server-side-rendering, but it's not replacing the correct state when it mounts. How do we ensure this happens?

**Comment by Akryum** [comment]
You need a solution to store the data resolved on the serve in the rendered HTML page so that the client can pick it up in the browser-side JS.

https://ssr.vuejs.org/guide/data.html#data-store
https://slides.com/akryum/vue-26-ssr-revolution#/21

**Comment by Akryum** [comment]
@yyx990803 I can't close the issue 😸

**Comment by DominusVilicus** [comment]
I figured out the issue. `Vue.observeable` doesn't trigger a re-render on server components when they import a child component which created the `Vue.observeable`

It's pretty complex, but can someone clarify whether components are re-rendered after `serverPrefetch`  updates a `Vue.observable` that multiple components rely on. It seems not to be happening in my case

**Comment by yyx990803** [comment]
Reactivity is entirely disabled during SSR, so there will be no "updates" - the state can be mutated, but it must happen before the render function is called.

**Comment by DominusVilicus** [comment]
This means that global state reactivity between different components is essentially disabled. Can we relook at this, or perhaps propose an RFC for 3.0?

**Comment by DominusVilicus** [comment]
It seems the render function for each component is being called before other components `serverPrefetch` is.

Shouldn't all `serverPrefetch`s for **all** components be called before any component is actually rendered?

**Comment by DominusVilicus** [comment]
is this expected behaviour @yyx990803, wouldn't it be more practical and expected for all the `serverPrefetch`s to be called before the render functions? especially when in the ssr vue examples state that the vuex store can be updated with serverPrefetch

https://ssr.vuejs.org/guide/data.html#data-store
https://ssr.vuejs.org/api/#serverprefetch

**Comment by Akryum** [comment]
> Shouldn't all serverPrefetchs for all components in the components: {...} be called before any component is actually rendered?

That is not possible, because children component instances are not created before render is called.

**Comment by yyx990803** [comment]
@DominusVilicus the code your originally posted is working as expected (per test case). By design, a component's `serverPrefetch` should only affect `data` of its own (or child components via passed props). The only case that it "doesn't work" is that when a component's `serverPrefetch` is mutating state that doesn't belong to itself.

**Comment by Akryum** [comment]
Consider a simple example where the component conditionally renders one component or another depending on its state:

```html
<template>
  <MyComponentA v-if="showA"/>
  <MyComponentB v-else/>
</template>
```

Now you can see we can't create the component instances of `MyComponentA` nor `MyComponentB` because we don't know which one will be created before actually rendering the current component.

**Comment by DominusVilicus** [comment]
Great points, I understand why now.

Evan is correct that the case doesn't work when using global state.

I guess this is why `Vuex` attaches itself to `root` component in `nuxtjs`, whereas my example attaches to `app` which is after `root` and `page/index.vue`. When passed to `root` it passes all the data to subcomponents

But why do the docs give examples of child components updating global state?

I totally understand your reasoning behind it though, just unsure how to proceed with my use-case

**Comment by hybridwebdev** [comment]
While i'm not well versed in SSR(it all seems like magic to me) I will interject this:

I believe any data collection SHOULD occur prior to a component being rendered
to me, both server side and client side...the model should be exactly the same
as soon as you provide inconsistencies, you provide surface area for bugs

**Comment by Akryum** [comment]
@DominusVilicus You should update the global state in a component which is a parent of all the components requiring it.

**Comment by Akryum** [comment]
@hybridwebdev You can't cover all the cases with that, for example components fetching their own data, like with Apollo GraphQL. 😸 But you can still do that, like before `serverPrefetch`, with `router.matchedComponents` (with the old limitations: only works on route-level components, no access to `this`).

**Comment by Akryum** [comment]
@DominusVilicus I don't think it's technically feasible without doing multiple re-render, which would be very bad performance-wise.

**Comment by DominusVilicus** [comment]
My idea behind it was compiling/creating all of the vnodes, then before stringifying/rendering them, calling all the serverPrefetch’s, which update the observeable, which should trigger the deps and update the vnode. After all the promises have been awaited, then finally render the components

**Comment by DominusVilicus** [comment]
![d5hwu7qQddLYno7Is2fZs1ZtPfU4832ZoOZ9gaTV](https://user-images.githubusercontent.com/17972275/56849675-9a562c80-6936-11e9-8c4f-17d6b1933583.gif)
Would you consider this optimal for a production scenario? probably not. (correct data is updated when the component mounts, but the server-side component should also have the correct data)

`code used (min-rep example)`
```vue
<template>
   <appLayout>
        hello {{ $state.me.firstName }}
    </appLayout>
</template>
```

I may be wrong, but isn't the goal of SSR to provide the same content on the client-side


**we all know why this is happening**, Vue doesn't have reactivity on the server-side, which is understandable, and totally fine and expected.

as @hybridwebdev mentioned:
> I believe any data collection SHOULD occur prior to a component being rendered
I think all data collection should occur prior to **any** component being rendered. Isn't the goal of SSR to provide and render the correct data and have parity with the client-side?

*yes, this only occurs when serverPrefetch is dealing with global component state*, but this is a common use-case.

I'd like to finalise with a final point. @Akryum's example of conditional components can be solved. It would just involve awaiting all `serverPrefetch()`s which notify/alert the relevant VNodes that depend on the data, and then finally rendering the VNodes

**Comment by Akryum** [comment]
> I think all data collection should occur prior to any component being rendered.

As I wrote in previous comments, this is achievable with `router.getMatchedComponents` which was the only way of data prefetching before `serverPrefetch`. Its functionality hasn't changed and you can still use it, although it has the same caveats/limitations as before (only route-level, no access to `this`).

> It would just involve awaiting all serverPrefetch()s promises

What if you have `serverPrefetch`s in the children components in the conditional in the example? You can't wait for all *possible* `serverPrefetch`s without progressively rendering the App first, since for a component A, the direct children component instances of A are only created when A is rendered.

> notify/alert the relevant VNodes that depend on the data, and then finally rendering the VNodes

Hum I think you may have a misunderstanding of what VNodes are. 😸 VNodes are created during render, not before. Basically VNodes are the parts that make up the Virtual DOM representation in memory created during render, when you call `h` (or the compiled template does).

> This way, every component is only rendered once

I think that's technically not possible.

> I'd be glad to implement

I've been thinking about the SSR data-prefetching problem for a year and came up with the current implementation of `serverPreftch`, but if you come up with a better implementation, it would be amazing!

Here are some slides about it: https://slides.com/akryum/vue-26-ssr-revolution#/21

**Comment by DominusVilicus** [comment]
Great diagram, it made me understand ssr further. I created another diagram (not the best) explaining how we could solve those caveats.

![image](https://user-images.githubusercontent.com/17972275/56851093-5ff68a80-694a-11e9-98af-1fb79d070ff2.png)


**this is how the proposal would work (pseudocode)**
```js
import index
import app
    createGlobalMixin() //see app.vue
import header

/**
  * this reads and compiles the template, creates vnodes and checks whether conditional components
  * should be compiled (as to not call serverPrefetch on `v-if="false"`s)
  * it finds <app> and `app.compile()`s it, which finds <header> and `header.compile()`s it too
  */
index.compile()
app.compile()
header.compile()

//all VNodes exist, and notify/deps are attached

index.serverPrefetch() //if provided (to add other page-specific local data state)
app.serverPrefetch() // adds some data to the $state object
header.serverPretch()

index.render() //recieves same global state as all components
app.render() //recieves same global state as all components
header.render() //recieves same global state as all components
```

I'll try to explain it in code

`app.js`
```js
import index from "page.vue"

let app = new Vue({
   render: h => h(page)
})}}

export default app //then the server creates client and server bundles
```

`page.vue`
```vue
<template>
    <app>
        <p>{{ foo }} (works and prints 'notBar') </p>
        <p>{{ $state.user.firstName }} would work in my proposal, but currently returns null</p>
    </app>
</template>
<script>
import app from "app.vue"

export default {
    data(){
        return {
            foo: 'bar'
        }
    },
    serverPrefetch(){
        this.foo = 'notBar'
    },
    components: {
        header
    }
}
</script>
```

`app.vue`
```vue
<template>
   <div>
       <header/>
       <slot/>
   </div>
</template>
<script>
import header from "header.vue"

let state = Vue.observable({
   user: {
        firstName: null,
        displayPicture: null
    }
})

/**
 * Vue.mixin applies a mixin globablly to every component
 */
Vue.mixin({
    beforeCreate(){
        this.$state = state
    }
})


export default {
    serverPrefetch(){
        let user = api.call('user')
        this.$state.user = user
    },
    components: {
        header
    }
}

</script>
```

`header.vue`
```vue
<template>
   <img :src="$state.user.displayPicture"> <!-- can access state both in my proposal, and current because it's rendered after app.render() -->
</template>
```

**Comment by Akryum** [comment]
VNodes aren't created until step 6 on your diagram (Rendering).

**Comment by Akryum** [comment]
> this reads and compiles the template

Templates are already compiled to `render` functions:
https://vuejs.org/v2/guide/render-function.html
https://template-explorer.vuejs.org

> creates vnodes

VNodes are created by `render()`

> checks whether conditional components should be compiled (as to not call serverPrefetch on `v-if="false"`s)

This happens when the component is rendered.

> all VNodes exist, and notify/deps are attached 

Again, no VNodes in a component exist until this component instance is rendered. Children component instances and their data/deps will be created when the component renders.

Example:

```
- A
   |- B
   |- C
```

What happens:

- A is created
- A is rendered
- B & C are created
- B & C are rendered

What cannot happen:

- A & B & C are created
- A & B & C are rendered

Since B & C are children of A, creating B & C requires rendering A first.

**Comment by DominusVilicus** [comment]
Yeah, I get you, my idea is to modify `render` internally so that it

1. render the vnodes
2. awaits ***all*** of the `serverPrefetch` promises (unlike only it's own)
3. and then finally render the vnodes to a string

**Comment by DominusVilicus** [comment]
This way there is no breaking changes, and serverPrefetch behaves more expectedly when behaving with global observeables

**Comment by Akryum** [comment]
```
- A
   |- B
   |- C
```

Ok, let's say you have `serverPrefetch` on both A, B & C, and that A, B & C all rely on different data from your API to correctly render. For example, A needs an API result to decide whether to render B or C with a `v-if`. B & C needs some other APIs calls to display data in their template too, and their `serverPrefetch` relies on some props passed by A.

What you are proposing will happen:

- A is created
- A renders
- Neither B nor C are created because A doesn't have the necessary data from the API to decide which one to render
- A's `serverPrefetch` is collected
- Await all `serverPrefetch`s, which means the one from A
- Generate the string representation of the app

There is a big issue here: neither B nor C where created, but you expected one of them to be created.

How to fix this? The only way is to render A again after the first `serverPrefetch`s awaiting, so that the VNode for B or C is created. So let's say we add a new rule: after the `serverPrefetch` is awaited, we re-render the component.

Here is what would happen:

- A is created
- A renders
- Neither B nor C are created because A doesn't have the necessary data from the API to decide which one to render
- A's `serverPrefetch` is collected
- Await all queued `serverPrefetch`s, which means the one from A
- A renders again
- B or C is created, let's say B (v-if was true)
- B renders
- B's `serverPrefech` is collected
- Await all queued `serverPrefetch`s, which means the one from B
- B renders again
- Generate the string representation of the app

We can already see we did multiple renders of the same component, which will affect performance. But it can get worse from there: what if B `serverPrefetch` actually mutated data that A depends on?

- A renders again
- A's `serverPrefetch` is collected
- Await all queued `serverPrefetch`s, which means the one from A
- A renders again
- B is updated (prop)
- B renders again
- etc

So now we actually rendered A four times and B three times! And this can go on and on and on...

---

### Takeaways

1 - It's impossible to have one and only one step with *all the possible* `serverPrefetch` in the app and call/await them all at once. The code can't guess if B or C would have been rendered and if we need to call their `serverPrefetch`. Also, if, like in the above example, B `serverPrefetch` relies on data passed via props by A, it can't work without having a intermediate step where A re-renders to pass the value via props to B.
2 - We can't render all the components in the app at once either after waiting for *all the possible* `serverPrefetch` (which is impossible, see takeaway 1).
3 - Mutating state that a parent component depends on is a *very bad idea* when doing SSR since it can lead to multiple cascading updates. You don't want the HTML page to take ages to be generated and sent to the client.
4 - We can't call/await `serverPrefetch` in a component instance after rendering it, otherwise we *have to* render it multiple times (and even call its `serverPrefetch` *again*).

**Comment by Akryum** [comment]
So basically your steps 2 and 4 in your diagram are not technically possible.

**Comment by Akryum** [comment]
> @DominusVilicus You should update the global state in a component which is a parent of all the components requiring it.

Let me rephrase it:

A component A should only mutate state that itself depends on OR state that its children component depend on. So for you use case, you should have a parent component which contains both `index`, `app` and `header` as children, and its `serverPrefetch` can mutate global state needed by `index`, `app` or `header`.

**Comment by DominusVilicus** [comment]
I've had long thought about this over the weekend and I understand the implications it would cause if child components caused updates. I think it may be worth considering adding the additional rerenders since they do change state, even if the child components do change it. I think this is what most vue users would like.

> what if **B** serverPrefetch actually mutated data that **A** depends on?

That's what is happening in the example image:
![56849675-9a562c80-6936-11e9-8c4f-17d6b1933583](https://user-images.githubusercontent.com/17972275/56884513-3011c900-6aa9-11e9-9d27-e3d33b90131e.gif)

1. `home.vue` is importing `appLayout.vue` and using it as a layout (slots, `<template><app></app></template>`)
2. `appLayout.vue` updates the global store which `home.vue` depends on (updates user in `serverPrefetch` so then user fetch isn't repeated in every single component, as all users of `appLayout.vue` depend on the user state, including itself and it's children)
3. as you can see, the `appLayout`'s child component `header` receives the correct data (`displayPicture`) since it is rendered afterwards
4. `home.vue` doesn't update it's VNodes based on the updated data (because it seems there's no reactivity, but there is sometimes like in `created()` afaik)

On the client-side, everything would work correctly because of reactivity, but in the server-side, this isn't the case, but I believe it should be. **Even if it involves rerenders**

I will continue to think of a viable solution to sort this issue that wouldn't add too much extra overhead.

I think this is what most developers using Vue would want. If they are causing a global state object to update in Vue, then of course they'd want it to be reflected across the entire app.

**Comment by EmilioAiolfi** [comment]
Guys, I have problems to replace asyncData for serverPrefetch. 
Hydrate doesn't happen. Show me a warning about the mismatch HTML the server for the client. 

I try to update the footer (in parent) component using getters to get data. This data is set for serverPrefetch in Details (in child) component.

AsyncData (this same method in HN example ) await all promises, But serverPrefetch doesn't the same approach.  

Please, show an example of this approach for update Parent component (with getters ) with serverPrefetch in the child with vuex

```
<App>
 <header>
 <router-view>
  <details>
</router-view>
<footer>
</App>
```

**Comment by doot0** [comment]
@EmilioAiolfi The mismatched HTML is commonly associated with authoring incorrect markup. Have you double checked that the markup in your components is valid?

**Comment by awacode21** [comment]
i guess i am dealing with the same problem. i need to build up a completely generic page coming from cms. dependant on that generic page (displayed in router view) the router-view is wrapped by a specific layout component (with for example header & footer).  So the child content sets the information which layout to choose.  When i understand this bug ticket here the right way, this is technically impossible? 
The last comment is 3 years ago. Are there any new facts about it?
i am working with vue 3, pinia, vite, ssr.

**Comment by Akryum** [comment]
@awacode21 As the app is rendered only once, you should probably rethink your app.

**Comment by awacode21** [comment]
@Akryum ok, i will rethink it. Thank you for your answer
